### PR TITLE
Stop retransmissions when we receive expected message.

### PR DESCRIPTION
**Pull request by sbernard31**
We should stop handshake message retransmission as soon as we receive the first message of the expected flight.
(see #228 for discussion about this)

### CODE REVIEW

**Comment by sbernard31** [comment]
I tested it with #305 by simulating some pskstore latency. It behaves much better.

**Comment by boaks** [comment]
RFC6347, 4.2.4

> 3. The implementation receives the next flight of messages: if this
> is the final flight of messages, the implementation transitions to
> FINISHED. If the implementation needs to send a new flight, it
> transitions to the PREPARING state. Partial reads (whether
> partial messages or only some of the messages in the flight) do
> not cause state transitions or timer resets.

I don't know, why "Partial reads (whether partial messages or only some of the messages in the flight) do not cause state transitions or timer resets." is specified, but before we merge this PR, it would be good to have some infos about that.

**Comment by sbernard31** [comment]
I will ask this on the tls@ietf.org mailing list but I'm not sure this can be applied to the letter to our implementation as we don't handle this exactly as the state machine described in this section.

**Comment by sbernard31** [comment]
Here a pertinent [answer](https://mailarchive.ietf.org/arch/msg/tls/oMQTOOHdujD03XDix9r4nTp59PM).

The problem is about the last flight.
A possible solution is to not stop retransmission on CHANGE CIPHER SPEC as last flight is always CHANGE CIPHER SPEC/ FINISHED ?

**Comment by boaks** [comment]
OK. 
LGTM.

Just as remark: if it's configurable, it would be easily possible to switch back in case of troubles.

**Comment by sbernard31** [comment]
I made the modification to not stop retransmissions on CIPHER SPEC CHANGE message.

About make it configurable, I feel this a bit strange.
If we really want to allow a workaround in case this could be a problem.
I prefer to allow  to override `getRecordLayerForPeer()` rather adding a new DTLS config parameters.

**Comment by boaks** [comment]
The idea of making it configurable is mainly triggered to allow to easily check both "strategies", early and late stops of retransmissions to see, which approach works better in a setup. I don't expect, that one works and the other not. I just expect, that the one has sometime advantages and sometimes the other. Overwritting would make it possible to change behaviour, but it's not that easy as adjust the configuration.
If `DTLSConnector` would have a flag, the new `cancelRetransmissions `could just use that to either call or not call `connection.cancelPendingFlight()`

**Comment by sbernard31** [comment]
> I just expect, that the one has sometime advantages and sometimes the other

I don't agree. I see the current PR as an optimization (bandwidth) and also a way to not disturb a not so stable DTLS implementation.
As we seem not so confident about this PR. I can understand you want a way to deactivate it, waiting we make sure this is the right way to do, but I hope we will not keep this forever.

So I will add a "temporary" config option to deactivate this features. (Hoping we will remove it before the 2.0.0 release)

**Comment by boaks** [comment]
(Hoping we will remove it before the 2.0.0 release)

agreed :-).

**Comment by sbernard31** [comment]
Config added.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java [hunk 1]
```diff
@@ -1641,7 +1641,10 @@ public void sendFlight(DTLSFlight flight) {
 			}
 
 			public void cancelRetransmissions() {
-				connection.cancelPendingFlight();
+				// TODO remove this check when this experimental feature will be
+				// not experimental anymore ^^
+				if (config.isEarlyStopRetransmission())
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: {
}
?

Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: Personally,  I prefer without brackets for "single line if", but if you prefer with it I can add it.

Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: I just prefer a "unique format" :-), 
therefore I think, with {} is, what is configured in the formatter.

Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: Good try about formatter, but this is not the case ;)

**Comment by boaks** [comment]
LGTM

**Comment by boaks** [comment]
@sbernard31 
Should I merge and pick it to master?

**Comment by sbernard31** [comment]
Integrated in 2.0.x branch (commit 660834f97d01813b9f5e4815b0f2e2a5a1e3c26d, f7a83d2bc3b3a8694a3e5e7cc1073f4965a1431e, d4ae205342d0bf7a16679a82ca811ce3e297e79a, 41180981f90d7ac441d5263a9c1efee5f72842f8)

**Comment by boaks** [comment]
@sbernard31 
Is this improvement intended for master also?

**Comment by sbernard31** [comment]
I don't know...
As it is "experimental" we can wait to confirm if it really works before to ask ourself if we back-port it in master ?

**Pull request closed without merging**